internal/pkg/service/common/etcdop/op/atomic_do_test.go

```diff
@@ -0,0 +1,134 @@
+package op_test
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/etcdhelper"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/etcdlogger"
+)
+
+func TestAtomicFromCtx_Misuse1(t *testing.T) {
+	t.Parallel()
+	assert.PanicsWithError(t, "no atomic operation found in the context", func() {
+		op.AtomicFromCtx(context.Background())
+	})
+}
+
+func TestAtomicFromCtx_Misuse2(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	client := etcdhelper.ClientForTest(t, etcdhelper.TmpNamespace(t))
+	atomicOp := op.Atomic(client, &op.NoResult{}).
+		Write(func(ctx context.Context) op.Op {
+			return etcdop.
+				NewKey("keys/key").Put(client, "value").
+				WithOnResult(func(result op.NoResult) {
+					op.AtomicFromCtx(ctx) // <<<<<<<<< panic
```

Comment:
Why this causes panic? The `Write` function has context within it could operate. Is it because we can have multiple reads and therefore the context has to be more generic than the `Write` one?

Reply:
The `on result` calback is invoked after the write transaction, so you cannot add some new operations to already invoked AtomicOp.

https://github.com/keboola/keboola-as-code/blob/7a99f5a9d36728982b730f2a90979a710c3caa86/internal/pkg/service/common/etcdop/op/atomic_do_test.go#L40-L43

It is clear now? What change do you suggest?

cc @jachym-tousek-keboola @hosekpeter